packages/financial-templates-lib/src/clients/InsuredBridgeL1Client.ts

```diff
@@ -0,0 +1,181 @@
+  }
+
+  async _initialSetup(fromBlock: number): Promise<void> {
+    const whitelistedTokenEvents = await this.bridgeAdmin.getPastEvents("WhitelistToken", { fromBlock });
+    for (const whitelistedTokenEvent of whitelistedTokenEvents) {
+      const l1Token = whitelistedTokenEvent.returnValues.l1Token;
+      this.bridgePools[l1Token] = (new this.l1Web3.eth.Contract(
+        this.bridgePoolAbi,
+        whitelistedTokenEvent.returnValues.bridgePool
+      ) as unknown) as BridgePoolWeb3;
+      this.relays[l1Token] = {};
+    }
+  }
+
+  async update(): Promise<void> {
+    // Define a config to bound the queries by.
+    const blockSearchConfig = {
+      fromBlock: this.firstBlockToSearch,
+      toBlock: this.endingBlockNumber || (await this.l1Web3.eth.getBlockNumber()),
+    };
+
+    if (!this._isClientInitialized()) await this._initialSetup(blockSearchConfig.fromBlock);
+
+    // Fetch event information
+    // TODO: consider optimizing this further. Right now it will make a series of sequential BlueBird calls for each pool.
+    for (const l1Token of Object.keys(this.bridgePools)) {
+      const [
+        depositRelayedEvents,
+        relaySpedUpEvents,
+        relayDisputedEvents,
+        relaySettledEvents,
+      ] = await BluebirdPromise.all([
+        this.bridgePools[l1Token].getPastEvents("DepositRelayed", blockSearchConfig),
+        this.bridgePools[l1Token].getPastEvents("RelaySpedUp", blockSearchConfig),
+        this.bridgePools[l1Token].getPastEvents("RelayDisputed", blockSearchConfig),
+        this.bridgePools[l1Token].getPastEvents("RelaySettled", blockSearchConfig),
+      ]);
+
+      for (const depositRelayedEvent of depositRelayedEvents) {
+        const relayData = {
+          depositId: Number(depositRelayedEvent.returnValues.depositId),
+          sender: depositRelayedEvent.returnValues.sender,
+          slowRelayer: depositRelayedEvent.returnValues.slowRelayer,
+          disputedSlowRelayers: [],
+          instantRelayer: ZERO_ADDRESS,
+          depositTimestamp: depositRelayedEvent.returnValues.depositTimestamp,
+          recipient: depositRelayedEvent.returnValues.recipient,
+          l1Token: depositRelayedEvent.returnValues.l1Token,
+          amount: depositRelayedEvent.returnValues.amount,
+          slowRelayFeePct: depositRelayedEvent.returnValues.slowRelayFeePct,
+          instantRelayFeePct: depositRelayedEvent.returnValues.instantRelayFeePct,
+          realizedLpFeePct: depositRelayedEvent.returnValues.realizedLpFeePct,
+          priceRequestAncillaryDataHash: depositRelayedEvent.returnValues.priceRequestAncillaryDataHash,
+          depositHash: depositRelayedEvent.returnValues.depositHash,
+          depositContract: depositRelayedEvent.returnValues.depositContract,
+          relayState: relayState.Pending,
+        };
+
+        // If the local data contains this deposit ID then this is a re-relay of a disputed relay. In this case, we need
```

Comment:
> If the local data contains this deposit ID then this is a re-relay of a disputed relay.

This assumption only holds if the `depositRelayedEvents` are sorted from oldest to newest right? Do you enforce this sorting anywhere?

Reply:
This data comes from the `getPastEvents` call which will have the ordering in which the events were created. I dont think any issue is introduced through by this having a diffrent ordering? how do you think this would impact if the order was wrong, seeing we use the `depositHash` as the key in this object?

Reply:
the `slowRelayer` of the last `event` in a set of duplicate `depositRelayedEvent` will be set as the `slowRelayer`. Normally it should be that the last `slowRelayer` is the non-disputed relayer, but this could be mixed up if the order is different?